src/util/FeedSwitch.sol

```diff
@@ -0,0 +1,120 @@
+contract FeedSwitch {
+    error NotGuardian();
+    error FeedDecimalsMismatch();
+    error MaturityInPast();
+    error MaturityPassed();
+
+    address public immutable guardian;
+    IChainlinkFeed public immutable initialFeed;
+    IChainlinkFeed public immutable beforeMaturityFeed;
+    IChainlinkFeed public immutable afterMaturityFeed;
+    uint256 public immutable maturity;
+    uint256 public immutable timelockPeriod;
+
+    IChainlinkFeed public feed;
+    IChainlinkFeed public previousFeed;
+
+    uint256 public switchCompletedAt;
+
+    event FeedSwitchInitiated(address indexed newFeed);
+
+    constructor(
+        address _feed,
+        address _beforeMaturityFeed,
+        address _afterMaturityFeed,
+        uint256 _timelockPeriod,
+        address _pendlePT,
+        address _guardian
+    ) {
+        feed = IChainlinkFeed(_feed);
+        initialFeed = IChainlinkFeed(_feed);
+        beforeMaturityFeed = IChainlinkFeed(_beforeMaturityFeed);
+        afterMaturityFeed = IChainlinkFeed(_afterMaturityFeed);
+        if (
+            beforeMaturityFeed.decimals() != 18 ||
+            afterMaturityFeed.decimals() != 18 ||
+            feed.decimals() != 18
+        ) revert FeedDecimalsMismatch();
+
+        timelockPeriod = _timelockPeriod;
+        maturity = IPendlePT(_pendlePT).expiry();
+        if (maturity < block.timestamp) revert MaturityInPast();
+
+        guardian = _guardian;
+    }
+
+    /// @notice Initiate the feed switch, entering the timelock period
+    /// @dev Can only be called by the guardian and can be done before maturity, after that, the feed will return afterMaturityFeed data
+    function initiateFeedSwitch() external {
+        if (msg.sender != guardian) revert NotGuardian();
+        if (block.timestamp > maturity) revert MaturityPassed();
+
+        if (switchCompletedAt < block.timestamp) {
+            switchCompletedAt = block.timestamp + timelockPeriod;
+        } else switchCompletedAt = 0;
+
+        if (address(feed) == address(initialFeed)) {
+            feed = beforeMaturityFeed;
+            previousFeed = initialFeed;
+        } else {
+            feed = initialFeed;
+            previousFeed = beforeMaturityFeed;
+        }
+
+        emit FeedSwitchInitiated(address(feed));
+    }
+
+    /// @notice Get the current feed data
+    /// @return roundId The round ID
+    /// @return price The price of the asset
+    /// @return startedAt The timestamp of the start of the round
+    /// @return updatedAt The timestamp of the last update
+    /// @return answeredInRound The round ID in which the price was answered
+    function latestRoundData()
+        public
+        view
+        returns (
+            uint80 roundId,
+            int256 price,
+            uint256 startedAt,
+            uint256 updatedAt,
+            uint80 answeredInRound
+        )
+    {
+        if (block.timestamp > maturity) {
```

Comment:
im not an expert on pendle, but does this match pendle logic? i.e. does maturity effect take place only *after* the timestamp, or on it?